File: packages/py/application/src/bolsa_application/operational_readiness.py
"""OR-6 — SEMI operational certification (ADR-035).

Cuatro estados discretos. Un FAIL crítico no se promedia a un %.
OE-1 PASS/FAIL/WARN sigue measure ≠ Accept. El carril AUTO no entra.
"""

from __future__ import annotations

from typing import Any, Literal

ReadinessState = Literal[
    "PAPER_READY",
    "PAPER_DEGRADED",
    "LIVE_EXPERIMENTAL",
    "LIVE_BLOCKED",
]
BrokerVenue = Literal["paper", "live"]
CtaKind = Literal["execute", "protect"]

_READINESS_RULE = (
    "no averaging · critical FAIL is not fifty-percent ready · measure ≠ Accept"
)


def _norm_venue(raw: str | None) -> BrokerVenue:
    venue = (raw or "paper").strip().lower()
    return "live" if venue == "live" else "paper"


def derive_operational_readiness(
    *,
    broker_venue: BrokerVenue | str | None = None,
    kill_switch_effective: bool = False,
    portfolio_reconciliation_status: str | None = "not_wired",
    live_reconciliation_status: str | None = None,
    semi_path_mark: str | None = "PASS",
    live_adapter_wired: bool | None = None,
) -> dict[str, Any]:
    """Mapea runtime + OE-1 a un estado. AUTO marks no son parámetro a propósito."""

    venue = _norm_venue(broker_venue)
    reasons: list[str] = []
    notes: list[str] = []

    recon = (portfolio_reconciliation_status or "not_wired").strip().lower()
    if recon == "drift":
        reasons.append("portfolio_drift")
    elif recon != "ok":
        reasons.append("recon_not_certified")

    if kill_switch_effective:
        reasons.append("kill_switch")

    semi = (semi_path_mark or "PASS").strip().upper()
    if semi == "UNAVAILABLE":
        reasons.append("semi_path_unavailable")
    elif semi == "WARN":
        notes.append("thin_semi_evidence")

    live = (live_reconciliation_status or "").strip().lower()
    if venue == "live":
        if live == "drift":
            reasons.append("live_drift")
        elif live == "unavailable":
            reasons.append("live_unavailable")
        if live_adapter_wired is False:
            reasons.append("live_adapter_not_wired")
        notes.append("live_not_accepted")

    if venue == "live":
        state: ReadinessState = (
            "LIVE_BLOCKED" if reasons else "LIVE_EXPERIMENTAL"
        )
    else:
        state = "PAPER_DEGRADED" if reasons else "PAPER_READY"

    return {
        "state": state,
        "venue": venue,
        "reasons": reasons,
        "notes": notes,
        "rule": _READINESS_RULE,
    }


def execute_cta_label(
    venue: BrokerVenue | str | None = None,
    *,
    kind: CtaKind = "execute",
) -> str:
    """Copy de firma. Protect no es envío a broker."""

    if kind == "protect":
        return "Confirmar protección"
    if _norm_venue(venue) == "live":
        return "Ejecutar en LIVE"
    return "Ejecutar en PAPER"

File: packages/shared/src/cognitive/operational-readiness.ts
/**
 * OR-6 — SEMI operational certification (ADR-035).
 * Cuatro estados discretos. Un FAIL crítico no se promedia a un %.
 * AUTO lane no entra. Espejo de operational_readiness.py.
 */

export type OperationalReadinessStateV1 =
  | "PAPER_READY"
  | "PAPER_DEGRADED"
  | "LIVE_EXPERIMENTAL"
  | "LIVE_BLOCKED";

export type ExecuteCtaKindV1 = "execute" | "protect";

type Venue = "paper" | "live";

export type OperationalReadinessInputV1 = {
  brokerVenue?: Venue | string | null;
  killSwitchEffective?: boolean;
  portfolioReconciliationStatus?: string | null;
  liveReconciliationStatus?: string | null;
  semiPathMark?: string | null;
  liveAdapterWired?: boolean | null;
};

export type OperationalReadinessV1 = {
  state: OperationalReadinessStateV1;
  venue: Venue;
  reasons: string[];
  notes: string[];
  rule: string;
};

const READINESS_RULE =
  "no averaging · critical FAIL is not fifty-percent ready · measure ≠ Accept";

function normVenue(raw: string | null | undefined): Venue {
  return String(raw ?? "paper")
    .trim()
    .toLowerCase() === "live"
    ? "live"
    : "paper";
}

export function deriveOperationalReadiness(
  input: OperationalReadinessInputV1 = {},
): OperationalReadinessV1 {
  const venue = normVenue(input.brokerVenue);
  const reasons: string[] = [];
  const notes: string[] = [];

  const recon = String(input.portfolioReconciliationStatus ?? "not_wired")
    .trim()
    .toLowerCase();
  if (recon === "drift") {
    reasons.push("portfolio_drift");
  } else if (recon !== "ok") {
    reasons.push("recon_not_certified");
  }

  if (input.killSwitchEffective) {
    reasons.push("kill_switch");
  }

  const semi = String(input.semiPathMark ?? "PASS")
    .trim()
    .toUpperCase();
  if (semi === "UNAVAILABLE") {
    reasons.push("semi_path_unavailable");
  } else if (semi === "WARN") {
    notes.push("thin_semi_evidence");
  }

  const live = String(input.liveReconciliationStatus ?? "")
    .trim()
    .toLowerCase();
  if (venue === "live") {
    if (live === "drift") {
      reasons.push("live_drift");
    } else if (live === "unavailable") {
      reasons.push("live_unavailable");
    }
    if (input.liveAdapterWired === false) {
      reasons.push("live_adapter_not_wired");
    }
    notes.push("live_not_accepted");
  }

  const state: OperationalReadinessStateV1 =
    venue === "live"
      ? reasons.length > 0
        ? "LIVE_BLOCKED"
        : "LIVE_EXPERIMENTAL"
      : reasons.length > 0
        ? "PAPER_DEGRADED"
        : "PAPER_READY";

  return {
    state,
    venue,
    reasons,
    notes,
    rule: READINESS_RULE,
  };
}

export function executeCtaLabel(
  venue?: Venue | string | null,
  kind: ExecuteCtaKindV1 = "execute",
): string {
  if (kind === "protect") return "Confirmar protección";
  return normVenue(venue) === "live" ? "Ejecutar en LIVE" : "Ejecutar en PAPER";
}
